src/core/tsi/alts/handshaker/alts_handshaker_client.h

```diff
@@ -34,131 +34,209 @@
-
-///
-/// This method handles handshaker response returned from ALTS handshaker
-/// service. Note that the only reason the API is exposed is that it is used in
-/// alts_shared_resources.cc.
-///
-///- client: an ALTS handshaker client instance.
-///- is_ok: a boolean value indicating if the handshaker response is ok to read.
-///
-void alts_handshaker_client_handle_response(alts_handshaker_client* client,
-                                            bool is_ok);
-
-// Returns the max number of concurrent handshakes that are permitted.
-//
-// Exposed for testing purposes only.
-size_t MaxNumberOfConcurrentHandshakes();
+class AltsHandshakerClient {
+ public:
+  ///
+  /// This factory method creates an ALTS handshaker client.
+  ///
+  ///- handshaker: ALTS TSI handshaker to which the created handshaker client
+  ///  belongs to.
+  ///- channel: grpc channel to ALTS handshaker service.
+  ///- handshaker_service_url: address of ALTS handshaker service in the format
+  /// of
+  ///  "host:port".
+  ///- interested_parties: set of pollsets interested in this connection.
+  ///- options: ALTS credentials options containing information passed from TSI
+  ///  caller (e.g., rpc protocol versions)
+  ///- target_name: the name of the endpoint that the channel is connecting to,
+  ///  and will be used for secure naming check
+  ///- grpc_cb: gRPC provided callbacks passed from TSI handshaker.
+  ///- cb: callback to be executed when tsi_handshaker_next API compltes.
+  ///- user_data: argument passed to cb.
+  ///- vtable_for_testing: ALTS handshaker client vtable instance used for
+  ///  testing purpose. ~~~~DEPRECIATED~~~~
+  ///- is_client: a boolean value indicating if the created handshaker client is
+  ///  used at the client (is_client = true) or server (is_client = false) side.
+  ///- max_frame_size: Maximum frame size used by frame protector (User
+  /// specified
+  ///  maximum frame size if present or default max frame size).
+  ///
+  /// It returns the created ALTS handshaker client on success, and NULL
+  /// on failure.
+  ///
+  static AltsHandshakerClient* alts_grpc_handshaker_client_create(
+      alts_tsi_handshaker* handshaker, grpc_channel* channel,
+      const char* handshaker_service_url, grpc_pollset_set* interested_parties,
+      grpc_alts_credentials_options* options, const grpc_slice& target_name,
+      grpc_iomgr_cb_func grpc_cb, tsi_handshaker_on_next_done_cb cb,
+      void* user_data, alts_handshaker_client_vtable* vtable_for_testing,
+      bool is_client, size_t max_frame_size, std::string* error);
+
+  ///
+  /// This method cancels previously scheduled, but yet executed handshaker
+  /// requests to ALTS handshaker service. After this operation, the handshake
+  /// will be shutdown, and no more handshaker requests will get scheduled.
+  ///
+  ///- client: ALTS handshaker client instance.
+  ///
+  static void alts_handshaker_client_shutdown(AltsHandshakerClient* client);
```

Comment:
I think we should change these from being static and self referentials, the following should be functionally identical.

```
AltsHandshakerClient* client = <somehow created>
alts_handshaker_client_shutdown(client);
client->Shutdown(); // (a new instance method)
```

Reply:
I was originally going to do that, but I realized that would involve changing a lot of code outside this class. Maybe I can open up another PR for this afterward. For this PR, I decided to keep those as sort of wrapper functions.

Reply:
On second thought, I realize that won't work because the static function won't be able to call private methods or access private variables anyways. Alright, I will just bite the bullet and reformat the other code as well.